### PR TITLE
feat(port): make corpses bleed similar from DDA

### LINKED ISSUE #4726: Add bleeding as a butchering option
### Is your feature request related to a problem? Please describe.

It makes sense to me that butchering should allow the user to pour the blood onto a container. On a quick, naive testing, it seems that only zombies provide this option:

image

Butchering a mi-go, wild boar, nor a rabbit does not provide that option.

### Describe the solution you'd like

I would like it if it also provides this option for other corpses, as this would be helpful for recipes that require blood. 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by chrsrns**
## Purpose of change

Fixes #4726 when done.

## Describe the solution

Provides a way to 'bleed' corpses, which is seen in DDA.

## Describe alternatives you've considered

Just add the bloods to the entries in `harvest.json` without providing the bleed feature.

## Testing

Nothing yet.

## Additional context

Eventually, this shows up after bleeding a corpse:
image

### For the feature itself:
- [x] Add the `Bleed corpse` entry to the UI.
- [x] Finish the handler for bleeding corpses.
- [x] Add the blood entries to `harvest.json`.
- [x] Add recipes that uses blood as an alternative ingredient.
- [x] Make sure that the blood draw kit can no longer draw blood from bled corpses.

### Other checklist:
- [ ] Document the changes in the appropriate location in the `doc/` folder.
- [ ] If documentation for this feature does not exist, please write it or at least note its lack in PR description.
- [ ] New localizable fields need to be added to the `lang/bn_extract_json_strings.sh` script if it does not support them yet.

### CODE REVIEW

**Comment by chrsrns** [comment]
Okay I don't get why that general build matrix test failed.

**Comment by chaosvolt** [comment]
Let's see, might be uh...
```
(~[slow] ~[.],starting_items)=> (continued from above) ERROR: The do_turn function ACT_BLEED doesn't correspond to a valid activity_type.
(~[slow] ~[.],starting_items)=> 16:20:47.086 ERROR DEBUGMSG : src/activity_type.cpp:98 [static void activity_type::check_consistency()] The finish_function ACT_BLEED doesn't correspond to a valid activity_type
```

**Comment by Lamandus** [comment]
what can you do with the blood at the moment?

**Comment by chrsrns** [comment]
> Let's see, might be uh...
> 
> ```
> (~[slow] ~[.],starting_items)=> (continued from above) ERROR: The do_turn function ACT_BLEED doesn't correspond to a valid activity_type.
> (~[slow] ~[.],starting_items)=> 16:20:47.086 ERROR DEBUGMSG : src/activity_type.cpp:98 [static void activity_type::check_consistency()] The finish_function ACT_BLEED doesn't correspond to a valid activity_type
> ```

Ohh. Found it. Totally missed that

**Comment by chrsrns** [comment]
> what can you do with the blood at the moment?

not much in the base game IFAIK, but mods like Arcana have some recipes that uses blood, and its a bit tedious to get a lot of those using the blood draw kit

**Comment by Lamandus** [comment]
> > what can you do with the blood at the moment?
> 
> not much in the base game IFAIK, but mods like Arcana have some recipes that uses blood, and its a bit tedious to get a lot of those using the blood draw kit

that is not good. Since we add these to the main game, there should be use for it


please consider the following:
1. add a recipe to make concentrated tainted blood via cooking. It should lower the volume by 10.
2. make concentrated blood be used in any serum recipe that uses tainted flesh and replace it
3. tainted tornado is using tainted meat, using the blood would fit more.

**Comment by chrsrns** [comment]
> > > what can you do with the blood at the moment?
> > 
> > 
> > not much in the base game IFAIK, but mods like Arcana have some recipes that uses blood, and its a bit tedious to get a lot of those using the blood draw kit
> 
> that is not good. Since we add these to the main game, there should be use for it
> 
> please consider the following:
> 
>     1. add a recipe to make concentrated tainted blood via cooking. It should lower the volume by 10.
> 
>     2. make concentrated blood be used in any serum recipe that uses tainted flesh and replace it
> 
>     3. tainted tornado is using tainted meat, using the blood would fit more.

that makes sense. will do

**Comment by chaosvolt** [comment]
Using them for mutagen seems like the most fitting vanilla use. There's also quest usage, I'll likely want to later on confirm that it interacts correctly with the blood draw kit, at minimum not letting you draw blood from bodies that've been bled dry, ideally also using the kit affecting how much blood can be harvested but I can see that being harder to implement...

**Comment by chaosvolt** [comment]
One thing I realized, it'll prompt the player with liquid handling every time they butcher, won't it?

**Comment by chrsrns** [comment]
> One thing I realized, it'll prompt the player with liquid handling every time they butcher, won't it?

```
                if( p.is_npc() || action != butcher_type::BLEED ) {
```

I thought this would prevent that for non-bleed actions. I only assumed tho didn't really test it. Will test later.

**Comment by chaosvolt** [comment]
I feel that's fine. Being able to bleed the corpse as a separate action means the blood's accessible, while it spilling if you do other butchery actions means it won't constantly pester the player about liquid handling, which would make butchering multiple corpses immensely annoying.

File: data/json/recipes/chem/mutagen.json [hunk 1]
```diff
@@ -244,7 +245,7 @@
     "using": [ [ "mutagen_production_standard", 25 ] ],
     "components": [
       [ [ "mutagen", 1 ] ],
-      [ [ "meat_tainted", 3 ], [ "dry_meat_tainted", 3 ] ],
+      [ [ "meat_tainted", 3 ], [ "dry_meat_tainted", 3 ], [ "blood_tainted_concentrate", 9 ] ],
```
Comment (Lamandus) [inline data/json/recipes/chem/mutagen.json:248]:
```suggestion
      [ [ "blood_tainted_concentrate", 9 ] ],
```
I would rather see only powder/fluids in making serums.

File: data/json/recipes/chem/mutagen.json [hunk 2]
```diff
@@ -276,7 +277,11 @@
     "batch_time_factors": [ 80, 20 ],
     "book_learn": [ [ "recipe_animal", 7 ] ],
     "using": [ [ "mutagen_production_standard", 25 ] ],
-    "components": [ [ [ "mutagen", 1 ] ], [ [ "meat", 3 ] ], [ [ "ammonia", 1 ], [ "lye_powder", 100 ] ] ],
+    "components": [
+      [ [ "mutagen", 1 ] ],
+      [ [ "meat", 3 ], [ "animal_blood_concentrate", 9 ] ],
```
Comment (Lamandus) [inline data/json/recipes/chem/mutagen.json:282]:
```suggestion
      [ [ "animal_blood_concentrate", 9 ] ],
```

File: data/json/recipes/chem/mutagen.json [hunk 3]
```diff
@@ -306,7 +311,11 @@
     "batch_time_factors": [ 80, 20 ],
     "book_learn": [ [ "recipe_animal", 8 ] ],
     "using": [ [ "mutagen_production_standard", 25 ] ],
-    "components": [ [ [ "mutagen", 1 ] ], [ [ "meat", 3 ] ], [ [ "ammonia", 1 ], [ "lye_powder", 100 ] ] ],
+    "components": [
+      [ [ "mutagen", 1 ] ],
+      [ [ "meat", 3 ], [ "animal_blood_concentrate", 9 ] ],
```
Comment (Lamandus) [inline data/json/recipes/chem/mutagen.json:316]:
```suggestion
      [ [ "animal_blood_concentrate", 9 ] ],
```

File: data/json/recipes/chem/mutagen.json [hunk 4]
```diff
@@ -336,7 +345,11 @@
     "batch_time_factors": [ 80, 20 ],
     "book_learn": [ [ "recipe_animal", 8 ] ],
     "using": [ [ "mutagen_production_standard", 25 ] ],
-    "components": [ [ [ "mutagen", 1 ] ], [ [ "meat", 1 ] ], [ [ "ammonia", 1 ], [ "lye_powder", 100 ] ] ],
+    "components": [
+      [ [ "mutagen", 1 ] ],
+      [ [ "meat", 1 ], [ "animal_blood_concentrate", 3 ] ],
```
Comment (Lamandus) [inline data/json/recipes/chem/mutagen.json:350]:
```suggestion
      [ [ "animal_blood_concentrate", 3 ] ],
```

File: data/json/recipes/chem/mutagen.json [hunk 5]
```diff
@@ -366,7 +379,11 @@
     "batch_time_factors": [ 80, 20 ],
     "book_learn": [ [ "recipe_animal", 8 ] ],
     "using": [ [ "mutagen_production_standard", 25 ] ],
-    "components": [ [ [ "mutagen", 1 ] ], [ [ "meat", 3 ] ], [ [ "ammonia", 1 ], [ "lye_powder", 100 ] ] ],
+    "components": [
+      [ [ "mutagen", 1 ] ],
+      [ [ "meat", 3 ], [ "animal_blood_concentrate", 9 ] ],
```
Comment (Lamandus) [inline data/json/recipes/chem/mutagen.json:384]:
```suggestion
      [ [ "animal_blood_concentrate", 9 ] ],
```

File: data/json/recipes/chem/mutagen.json [hunk 6]
```diff
@@ -396,7 +413,11 @@
...
```
Comment (Lamandus) [inline data/json/recipes/chem/mutagen.json:418]:
```suggestion
      [ [ "animal_blood_concentrate", 9 ] ],
```

File: data/json/recipes/chem/mutagen.json [hunk 7]
```diff
@@ -426,7 +447,11 @@
...
```
Comment (Lamandus) [inline data/json/recipes/chem/mutagen.json:452]:
```suggestion
      [ [ "animal_blood_concentrate", 9 ] ],
```

File: data/json/recipes/chem/mutagen.json [hunk 8]
```diff
@@ -550,7 +581,11 @@
     "batch_time_factors": [ 80, 20 ],
     "book_learn": [ [ "recipe_maiar", 8 ] ],
     "using": [ [ "mutagen_production_standard", 25 ] ],
-    "components": [ [ [ "mutagen", 1 ] ], [ [ "meat", 3 ] ], [ [ "ammonia", 1 ], [ "lye_powder", 100 ] ] ],
+    "components": [
+      [ [ "mutagen", 1 ] ],
+      [ [ "meat", 3 ], [ "animal_blood_concentrate", 9 ] ],
```
Comment (Lamandus) [inline data/json/recipes/chem/mutagen.json:586]:
```suggestion
      [ [ "animal_blood_concentrate", 9 ] ],
```

File: data/json/recipes/chem/mutagen.json [hunk 9]
```diff
@@ -458,7 +483,13 @@
     "using": [ [ "mutagen_production_standard", 25 ] ],
     "components": [
       [ [ "mutagen", 1 ] ],
-      [ [ "meat_tainted", 3 ], [ "dry_meat_tainted", 3 ], [ "can_chowder", 1 ], [ "can_clams", 1 ] ],
+      [
+        [ "meat_tainted", 3 ],
+        [ "dry_meat_tainted", 3 ],
+        [ "blood_tainted_concentrate", 9 ],
```
Comment (Lamandus) [inline data/json/recipes/chem/mutagen.json:489]:
```suggestion
       [ [ "can_chowder", 1 ], [ "can_clams", 1 ], [ "blood_tainted_concentrate", 9 ] ],
```

File: data/json/recipes/chem/mutagen.json [hunk 10]
```diff
@@ -458,7 +483,13 @@
     "using": [ [ "mutagen_production_standard", 25 ] ],
     "components": [
       [ [ "mutagen", 1 ] ],
-      [ [ "meat_tainted", 3 ], [ "dry_meat_tainted", 3 ], [ "can_chowder", 1 ], [ "can_clams", 1 ] ],
+      [
```
Comment (Lamandus) [inline data/json/recipes/chem/mutagen.json:486]:
```suggestion
```

File: data/json/recipes/chem/mutagen.json [hunk 11]
```diff
@@ -458,7 +483,13 @@
...
+        [ "meat_tainted", 3 ],
```
Comment (Lamandus) [inline data/json/recipes/chem/mutagen.json:487]:
```suggestion
```

File: data/json/recipes/chem/mutagen.json [hunk 12]
```diff
@@ -458,7 +483,13 @@
...
+        [ "dry_meat_tainted", 3 ],
```
Comment (Lamandus) [inline data/json/recipes/chem/mutagen.json:488]:
```suggestion
```

File: data/json/recipes/chem/mutagen.json [hunk 13]
```diff
@@ -458,7 +483,13 @@
...
+        [ "blood_tainted_concentrate", 9 ],
+        [ "can_chowder", 1 ],
+        [ "can_clams", 1 ]
+      ],
```
Comment (Lamandus) [inline data/json/recipes/chem/mutagen.json:492]:
```suggestion
```

File: data/json/recipes/chem/mutagen.json [hunk 14]
```diff
@@ -458,7 +483,13 @@
     "using": [ [ "mutagen_production_standard", 25 ] ],
     "components": [
       [ [ "mutagen", 1 ] ],
-      [ [ "meat_tainted", 3 ], [ "dry_meat_tainted", 3 ], [ "can_chowder", 1 ], [ "can_clams", 1 ] ],
+      [
+        [ "meat_tainted", 3 ],
+        [ "dry_meat_tainted", 3 ],
+        [ "blood_tainted_concentrate", 9 ],
+        [ "can_chowder", 1 ],
+        [ "can_clams", 1 ]
```
Comment (Lamandus) [inline data/json/recipes/chem/mutagen.json:491]:
```suggestion
```

File: data/json/recipes/chem/mutagen.json [hunk 15]
```diff
@@ -458,7 +483,13 @@
     "using": [ [ "mutagen_production_standard", 25 ] ],
     "components": [
       [ [ "mutagen", 1 ] ],
-      [ [ "meat_tainted", 3 ], [ "dry_meat_tainted", 3 ], [ "can_chowder", 1 ], [ "can_clams", 1 ] ],
+      [
+        [ "meat_tainted", 3 ],
+        [ "dry_meat_tainted", 3 ],
+        [ "blood_tainted_concentrate", 9 ],
+        [ "can_chowder", 1 ],
```
Comment (Lamandus) [inline data/json/recipes/chem/mutagen.json:490]:
```suggestion
```

**Review by Lamandus — CHANGES_REQUESTED** [review]
consider removing meat from serum making completely

**Comment by chrsrns** [comment]
I am pondering on recipe would make concentrated blood. The idea is that this just removes most of the water from it, so just heating it up and stirring continuously to prevent most of the clotting makes sense to me. I think this would just translate in game to needing a boiling 1 tool and heat. A part of me thinks that this would need an anticoagulant of sorts to make sense, but that is just a feeling, as I can't find any concrete basis online on how to concentrate blood.

Just removing concentrated blood and just using the blood itself is an option, but the reduced volume would be nice. Unnecessary at that point of the game, but nice.

My search history for this is wild.

**Comment by Lamandus** [comment]
> I am pondering on the recipe that would make concentrated blood. The idea is that this just removes most of the water from it, so just heating it up and stirring continuously to prevent most of the clotting makes sense to me. I think this would just translate in game to needing a boiling 1 tool and heat. A part of me thinks that this would need an anticoagulant of sorts to make sense, but that is just a feeling, as I can't find any concrete basis online on how to concentrate blood.
> 
> Just removing concentrated blood and just using the blood itself is an option, but the reduced volume would be nice. Unnecessary at that point of the game, but nice.
> 
> My search history for this is wild.

while not wholly correct, Aspirin (ASS) is an antiplatelet medication.
Alternative would be the use of a rag to get the chunks out (I do it usually when making blood-sausage)

speaking of which, can you make a "Blutwurst" recipe?

https://www.chefkoch.de/rezepte/2324751370527261/Die-Herstellung-von-Blutwurst.html

https://www-chefkoch-de.translate.goog/rezepte/2324751370527261/Die-Herstellung-von-Blutwurst.html?_x_tr_sl=de&_x_tr_tl=en&_x_tr_hl=de&_x_tr_pto=wapp (translated)

Blood sausage is very popular in Germany, so it would be a good use for animal blood.

Sausage casting, concentrated animal blood, chunks of meat and some salt/spices are what we use in Germany.

You can either add a new recipe, or put the concentrated blood into the wasteland sausage recipe. Your call.

**Comment by Relgar** [comment]
This is an exciting pull request that has many potential applications.

There are a fair few pieces of lab equipment that would benefit being given a 'use' or tool quality that works with bloods or liquids. They could be used to extract specific things from the many lifeforms we have that bleed: animals, humans, zombies, mi-go, nether creatures, or mutants. It could also analyse if specific useful compounds are present in blood. In turn, extracting these specific minerals or compounds from the above list can lead to applications such as: 

- enhanced healing items (IV bags FINALLY) 
- experimental mutagen serums instead of requiring tainted meat as mentioned
- maybe solve the zombie bite problem with a vaccine
- powering mi-go weapons with fuel extracted from mi-go blood, which would play nicely with the idea of biological weapons
- cooking / brewing / formulating many other foods / drinks /drugs with these

**Comment by chrsrns** [comment]
> You can either add a new recipe, or put the concentrated blood into the wasteland sausage recipe. Your call.

I went with creating a blutwurst item and recipe. Since it is a famous dish, it felt weird for it to be alike to the wateland sausage with negative morale.

**Comment by chrsrns** [comment]
> ## Purpose of change
> 
> Fixes #4726 when done.
> ## Describe the solution
> 
> Provides a way to 'bleed' corpses, which is seen in DDA.
> ## Describe alternatives you've considered
> 
> Just add the bloods to the entries in `harvest.json` without providing the bleed feature.
> ## Testing
> 
> Nothing yet.
> ## Additional context
> 
> Eventually, this shows up after bleeding a corpse: image
> ## Checklist
> ### For the feature itself:
> 
>     * [x]  Add the `Bleed corpse` entry to the UI.
> 
>     * [x]  Finish the handler for bleeding corpses.
> 
>     * [x]  Add the blood entries to `harvest.json`.
> 
>     * [ ]  Add recipes that uses blood as an alternative ingredient.
> 
>     * [ ]  Make sure that the blood draw kit can no longer draw blood from bled corpses.
> 
> 
> ### Other checklist:
> 
>     * [ ]  Document the changes in the appropriate location in the `doc/` folder.
> 
>     * [ ]  If documentation for this feature does not exist, please write it or at least note its lack in PR description.
> 
>     * [ ]  New localizable fields need to be added to the `lang/bn_extract_json_strings.sh` script if it does not support them yet.


    * [x]  Add recipes that uses blood as an alternative ingredient.

    * [x]  Make sure that the blood draw kit can no longer draw blood from bled corpses.

**Comment by scarf005** [comment]
image

while related to PR, it seems like those new recipes could've been a separate PR (so it's easier to review), could they get extracted out?

**Comment by chrsrns** [comment]
> while related to PR, it seems like those new recipes could've been a separate PR (so it's easier to review), could they get extracted out?

That's fair. Will do that now.

**Comment by chaosvolt** [comment]
Fixed a smol case of a harvest entry having blood defined twice and added human blood to some human harvest entries missing it. Looks good at first glance but not currently home so won't be able to compile and playtest just yet.

I'd be very tempted to ponder reducing the mass ratios of blood a lil but it depends on how much blood you can get out of a body, plus with all the things we've foisted on you already I figure it's something entirely fine to leave as something I can tinker with myself in a followup PR.

**Comment by chrsrns** [comment]
> plus with all the things we've foisted on you already I figure it's something entirely fine to leave as something I can tinker with myself in a followup PR.

I'm very new to the codebase so that's totally fine

**Review by chaosvolt — APPROVED** [review]
1. Compiled and load-tested.
2. Played around with bleeding and butchering of zeds and NPCs to test behavior.

Fixed some oddities I encountered during testing:
1. Fixed the way blood is zeroed out by flag checks so that it doesn't require all of the flags, thus fixing  it still dropping blood on dissection and the like.
2. Set the bleed action to grab specifically whatever is defined as a blood harvest entry instead of specifically searching for liquids in the harvest.
3. Removed the extra effect of a failed field dressing on bleeding since field dressing in general already automatically zeroes out the blood yield in an earier check.
4. Tweaked reduction of corpse weight to lose 10% of weight instead of 20% since havest entries all use 10% as the mass ratio.
5. Fixed incorrect message when trying to use blood draw kit on an already-bled corpse.
6. Edited comments in a few spots.

I found a bug where it's possible to dupe organs, that plus it always saying you fail to obtain yields you already harvested in an earlier action are two issues that predate this PR and I think I might fuck with it as a direct followup of this.

**Pull request merged**